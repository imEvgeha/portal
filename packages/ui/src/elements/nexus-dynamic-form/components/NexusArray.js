import React, {useContext, useState, useEffect} from 'react';
import PropTypes from 'prop-types';
import {Field as AKField} from '@atlaskit/form';
import {default as AKForm} from '@atlaskit/form/Form';
import EditorCloseIcon from '@atlaskit/icon/glyph/editor/close';
import {Button} from '@portal/portal-components';
import {get} from 'lodash';
import {useDispatch} from 'react-redux';
import {addToast} from '../../../toast/NexusToastNotificationActions';
import {NexusModalContext} from '../../nexus-modal/NexusModal';
import {DELETE, CANCEL, REMOVE_TITLE} from '../../nexus-tag/constants';
import {
    buildSection,
    checkFieldDependencies,
    getFieldConfig,
    getValidationFunction,
    renderNexusField,
    renderError,
    renderLabel,
} from '../utils';
import {VIEWS, DELETE_POPUP} from '../constants';
import './NexusArray.scss';

const NexusArray = ({
    name,
    view,
    data,
    fields,
    getValues,
    setFieldValue,
    confirmationContent,
    isRequired,
    isReadOnly,
    tooltip,
    dependencies,
    path,
    validationError,
    validation,
    selectValues,
    isUpdate,
    config,
    isEditable,
    generateMsvIds,
    sectionID,
}) => {
    const dispatch = useDispatch();
    const {openModal, closeModal} = useContext(NexusModalContext);
    // allData includes initialData and rows added/removed
    const [allData, setAllData] = useState(data);

    useEffect(() => {
        setAllData(data);
    }, [data, isUpdate]);

    const renderAddButton = () => {
        return (
            <Button
                label={`+ Add ${name}`}
                onClick={openEditModal}
                className="p-button-outlined p-button-secondary nexus-c-dynamic-form__add-button"
            />
        );
    };

    const onRemove = index => {
        const values = getValues();
        const editedData =
            path &&
            values[path].filter((obj, i) => {
                if (i === index && obj.selected) {
                    dispatch(
                        addToast({
                            detail: 'Selected territory cannot be removed',
                            severity: 'error',
                        })
                    );
                    return true;
                }

                return i !== index;
            });

        setAllData(editedData);
        setFieldValue(path, editedData);
        closeModal && closeModal();
    };

    const handleRemove = index => {
        if (confirmationContent) {
            const actions = [
                {
                    text: CANCEL,
                    onClick: closeModal,
                    appearance: 'default',
                },
                {
                    text: DELETE,
                    onClick: () => {
                        onRemove(index);
                        closeModal();
                    },
                    appearance: 'danger',
                },
            ];
            openModal(`${DELETE_POPUP} ${confirmationContent}`, {title: REMOVE_TITLE, width: 'medium', actions});
        } else {
            onRemove(index);
        }
    };

    const renderObject = (object, index) => {
        return (
            <div key={index} className="nexus-c-array__object">
                {view !== VIEWS.VIEW && (
                    <div className="nexus-c-array__remove-button" onClick={() => handleRemove(index)}>
                        <EditorCloseIcon size="medium" />
                    </div>
                )}
                <div className="row">{buildObject(fields, allData[index] || {}, index)}</div>
            </div>
        );
    };

    const buildObject = (fields = {}, initialData, index) => {
        return (
            <>
                {Object.keys(fields).map(key => {
                    return (
                        !getFieldConfig(fields[key], 'hidden', view) && (
                            <div key={`nexus-c-array__field ${key}`} className="col-6 array-field-wrapper">
                                {renderNexusField(`${path}[${index}].${key}`, view, getValues, generateMsvIds, {
                                    initialData,
                                    field: fields[key],
                                    sectionID: `${sectionID}.${index}`,
                                    selectValues,
                                    setFieldValue,
                                    config,
                                    shouldStackLabel: true,
                                })}
                            </div>
                        )
                    );
                })}
            </>
        );
    };

    const buildButtons = (dirty, submitting, reset) => {
        return (
            <div className="nexus-c-array__modal-buttons">
                <Button
                    label="Confirm"
                    type="submit"
                    className="p-button-outlined nexus-c-array__submit-button"
                    disabled={!dirty || submitting}
                />
                <Button
                    label="Cancel"
                    className="p-button-outlined p-button-secondary nexus-c-array__cancel-button"
                    onClick={() => {
                        reset();
                        closeModal();
                    }}
                />
            </div>
        );
    };

    const handleOnSubmit = values => {
        const createFormData = getValues();
        const arrayData = get(createFormData, path) ? get(createFormData, path) : [];
        // including the new row
        const editedArray = [...arrayData, values];
        setAllData(editedArray);
        setFieldValue(path, editedArray);
        closeModal();
    };

    const modalContent = () => {
        return (
            <div className="nexus-c-array__modal">
                <AKForm onSubmit={values => handleOnSubmit(values)}>
                    {({formProps, dirty, submitting, reset, getValues, setFieldValue}) => (
                        <form {...formProps}>
                            <div className="nexus-c-array__modal-fields">
                                {buildSection(fields, getValues, VIEWS.CREATE, null, null, {
                                    selectValues,
                                    setFieldValue,
                                    sectionID,
                                })}
                            </div>
                            {buildButtons(dirty, submitting, reset)}
                        </form>
                    )}
                </AKForm>
            </div>
        );
    };

    const openEditModal = e => {
        e.preventDefault();
        openModal(modalContent(), {
            title: <div className="nexus-c-array__modal-title">{`Add ${name} Data`}</div>,
            width: 'medium',
        });
    };

    const required = !!(
        checkFieldDependencies('required', view, dependencies, {formData: getValues(), config, isEditable}) ||
        isRequired
    );

    const readOnly = !!(
        checkFieldDependencies('readOnly', view, dependencies, {formData: getValues(), config, isEditable}) ||
        isReadOnly
    );
    return (
        <div className={`nexus-c-array ${validationError ? 'nexus-c-array--error' : ''}`}>
            <AKField
                name={path}
                isRequired={required}
                defaultValue={data}
                validate={value => getValidationFunction(value, validation, {type: 'array', isRequired: required})}
            >
                {({fieldProps, error}) => (
                    <>
                        {renderLabel(name, required, tooltip)}
                        {error && renderError(error)}
                        {validationError && <div>{validationError}</div>}
                    </>
                )}
            </AKField>
            {!readOnly && <div className="nexus-c-array__add">{view !== VIEWS.VIEW && renderAddButton()}</div>}
            <div className="nexus-c-array__objects">
                {allData && Array.isArray(allData) && allData.map((o, index) => renderObject(o, index))}
            </div>
        </div>
    );
};

NexusArray.propTypes = {
    name: PropTypes.string.isRequired,
    path: PropTypes.string,
    view: PropTypes.string,
    tooltip: PropTypes.string,
    data: PropTypes.array,
    fields: PropTypes.object,
    getValues: PropTypes.func,
    setFieldValue: PropTypes.func,
    confirmationContent: PropTypes.string,
    isRequired: PropTypes.bool,
    isReadOnly: PropTypes.bool,
    validationError: PropTypes.string,
    validation: PropTypes.array,
    dependencies: PropTypes.array,
    selectValues: PropTypes.object,
    isUpdate: PropTypes.bool,
    config: PropTypes.array,
    isEditable: PropTypes.bool,
    generateMsvIds: PropTypes.func,
    sectionID: PropTypes.string,
};

NexusArray.defaultProps = {
    path: '',
    view: VIEWS.VIEW,
    tooltip: null,
    data: [],
    fields: {},
    getValues: undefined,
    setFieldValue: undefined,
    confirmationContent: null,
    isRequired: false,
    isReadOnly: false,
    validationError: null,
    validation: [],
    dependencies: [],
    selectValues: {},
    isUpdate: false,
    config: [],
    isEditable: false,
    generateMsvIds: undefined,
    sectionID: '',
};

export default NexusArray;
