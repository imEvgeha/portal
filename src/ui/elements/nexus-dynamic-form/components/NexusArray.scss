@import '../../../../styles/atlaskit/mixins';
@import '../../../../styles/atlaskit/colors';

.nexus-c-array {
    padding: 0.75rem 1.25rem;
    margin-bottom: 5px;
    .nexus-c-array__add {
        text-align: right;
    }
    .nexus-c-array__objects {
        display: flex;
        flex-wrap: wrap;
        justify-content: space-between;
        align-content: space-between;

        > div {
            flex: 0 49.5%;
            border: 1px solid map-get($light-neutral-colors, light-neutral-1);
            box-sizing: border-box;
        }
    }
    .nexus-c-array__object {
        padding: 25px;
        margin-top: 15px;
    }
}

.nexus-c-array__modal {
    .nexus-c-array__modal-buttons {
        text-align: right;
    }
}
