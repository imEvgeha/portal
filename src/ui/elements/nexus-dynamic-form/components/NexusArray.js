import React, {useContext, useState} from 'react';
import PropTypes from 'prop-types';
import Button from '@atlaskit/button';
import {default as AKForm} from '@atlaskit/form/Form';
import EditorCloseIcon from '@atlaskit/icon/glyph/editor/close';
import {get} from 'lodash';
import {NexusModalContext} from '../../nexus-modal/NexusModal';
import {CANCEL, DELETE, REMOVE_TITLE} from '../../nexus-tag/constants';
import {buildSection, getFieldConfig, renderNexusField, getProperValues} from '../utils';
import {VIEWS, DELETE_POPUP} from '../constants';
import './NexusArray.scss';

const NexusArray = ({
    name,
    view,
    data,
    fields,
    getValues,
    setFieldValue,
    schema,
    setDisableSubmit,
    confirmationContent,
}) => {
    const {openModal, closeModal} = useContext(NexusModalContext);
    // allData includes initialData and rows added
    const [allData, setAllData] = useState(data);

    const renderAddButton = () => {
        return (
            <Button onClick={openEditModal} className="nexus-c-dynamic-form__add-button">
                {`+ Add ${name}`}
            </Button>
        );
    };

    const onRemove = index => {
        console.log(index);
    };

    const handleRemove = index => {
        if (confirmationContent) {
            const actions = [
                {
                    text: CANCEL,
                    onClick: closeModal,
                    appearance: 'default',
                },
                {
                    text: DELETE,
                    onClick: () => {
                        onRemove(index);
                        closeModal();
                    },
                    appearance: 'danger',
                },
            ];
            openModal(`${DELETE_POPUP} ${confirmationContent}`, {title: REMOVE_TITLE, width: 'medium', actions});
        } else {
            onRemove(index);
        }
    };

    const renderObject = (object, index) => {
        return (
            <div key={index} className="nexus-c-array__object">
                <div className="nexus-c-tag__remove-button" onClick={() => handleRemove(index)}>
                    <EditorCloseIcon size="medium" />
                </div>
                {buildObject(fields, allData[index] || {}, index)}
            </div>
        );
    };

    const buildObject = (fields = {}, initialData, index) => {
        return (
            <>
                {Object.keys(fields).map(key => {
                    return (
                        !getFieldConfig(fields[key], 'hidden', view) && (
                            <div
                                key={`nexus-c-array__field ${key}`}
                                className={`nexus-c-array__field ${fields[key].className ? fields[key].className : ''}`}
                            >
                                {renderNexusField(
                                    `${name}[${index}].${key}`,
                                    view,
                                    getValues,
                                    initialData,
                                    fields[key]
                                )}
                            </div>
                        )
                    );
                })}
            </>
        );
    };

    const buildButtons = (dirty, submitting, reset) => {
        return (
            <div className="nexus-c-array__modal-buttons">
                <Button
                    type="submit"
                    className="nexus-c-array__submit-button"
                    appearance="primary"
                    isDisabled={!dirty || submitting}
                >
                    Confirm
                </Button>
                <Button
                    className="nexus-c-array__cancel-button"
                    onClick={() => {
                        reset();
                        closeModal();
                    }}
                >
                    Cancel
                </Button>
            </div>
        );
    };

    const handleOnSubmit = values => {
        const formData = getValues();
        const arrayData = get(formData, name) ? get(formData, name) : [];
        // including the new row
        const editedArray = [...arrayData, values];

        const properValues = editedArray.map(v => getProperValues(schema, v));

        setAllData(editedArray);
        setFieldValue(name, properValues);
        setDisableSubmit(false);
        closeModal();
    };

    const modalContent = () => {
        return (
            <div className="nexus-c-array__modal">
                <AKForm onSubmit={values => handleOnSubmit(values)}>
                    {({formProps, dirty, submitting, reset, getValues}) => (
                        <form {...formProps}>
                            {buildSection(fields, getValues, VIEWS.CREATE, {})}
                            {buildButtons(dirty, submitting, reset)}
                        </form>
                    )}
                </AKForm>
            </div>
        );
    };

    const openEditModal = () => {
        openModal(modalContent(), {
            title: <div className="nexus-c-array__modal-title">{`Add ${name} Data`}</div>,
            width: 'medium',
        });
    };

    return (
        <div className="nexus-c-array">
            <div className="nexus-c-array__add">{view !== VIEWS.VIEW && renderAddButton()}</div>
            <div className="nexus-c-array__objects">{allData.map((o, index) => renderObject(o, index))}</div>
        </div>
    );
};

NexusArray.propTypes = {
    schema: PropTypes.array.isRequired,
    name: PropTypes.string.isRequired,
    view: PropTypes.string,
    data: PropTypes.array,
    fields: PropTypes.object,
    getValues: PropTypes.func,
    setFieldValue: PropTypes.func,
    setDisableSubmit: PropTypes.func,
    confirmationContent: PropTypes.string,
};

NexusArray.defaultProps = {
    view: VIEWS.VIEW,
    data: [],
    fields: {},
    getValues: undefined,
    setFieldValue: undefined,
    setDisableSubmit: undefined,
    confirmationContent: null,
};

export default NexusArray;
