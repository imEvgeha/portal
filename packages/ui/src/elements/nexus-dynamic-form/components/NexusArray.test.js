import React from 'react';
import {shallow} from 'enzyme';
import NexusArray from './NexusArray';

describe('NexusArray', () => {
    let wrapper = null;
    let setFieldValueMock;
    let closeModalMock;

    const fields = {
        selected: {
            name: 'Selected',
            type: 'boolean',
            path: 'territory.selected',
            isReadOnly: true,
        },
        country: {
            name: 'Country',
            type: 'string',
            path: 'territory.country',
        },
    };

    const data = [
        {
            country: 'IL',
            selected: false,
            dateSelected: null,
            rightContractStatus: 'Pending',
            vuContractId: [],
            hide: null,
            comment: 'yyyyy',
            dateWithdrawn: null,
        },
        {
            country: 'US',
            selected: false,
            dateSelected: null,
            rightContractStatus: 'MatchedOnce',
            vuContractId: [],
            hide: null,
            comment: '',
            dateWithdrawn: null,
        },
    ];

    beforeEach(() => {
        setFieldValueMock = jest.fn();
        closeModalMock = jest.fn();
        const props = {
            name: 'territory',
            view: 'EDIT',
            data,
            fields,
            defaultValue: 'rght_zrp8g',
            getValues: () => {},
            setFieldValue: setFieldValueMock,
            closeModal: closeModalMock,
        };

        wrapper = shallow(<NexusArray {...props} />);
    });

    it('should match snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('should render an add button when is in edit mode', () => {
        const nexusArray = wrapper.find('.nexus-c-array');
        expect(nexusArray.length).toEqual(1);
        const addBtn = wrapper.find('.nexus-c-array__add');
        expect(addBtn.length).toEqual(1);
    });

    it('should render 2 object with 2 fields each', () => {
        const nexusArrayObjects = wrapper.find('.nexus-c-array__object');
        expect(nexusArrayObjects.length).toEqual(2);

        const nexusArrayFields = wrapper.find('.array-field-wrapper');
        expect(nexusArrayFields.length).toEqual(4);
    });

    it('should render 2 remove buttons and onClick the field value should be updated', () => {
        const nexusRemoveButtons = wrapper.find('.nexus-c-array__remove-button');
        expect(nexusRemoveButtons.length).toEqual(2);
        nexusRemoveButtons.at(0).simulate('click');
        expect(setFieldValueMock.mock.calls.length).toEqual(1);
    });
});
