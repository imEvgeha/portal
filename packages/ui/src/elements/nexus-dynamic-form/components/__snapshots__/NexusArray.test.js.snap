// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NexusArray should match snapshot 1`] = `
<div
  className="nexus-c-array "
>
  <Field
    defaultValue={
      Array [
        Object {
          "comment": "yyyyy",
          "country": "IL",
          "dateSelected": null,
          "dateWithdrawn": null,
          "hide": null,
          "rightContractStatus": "Pending",
          "selected": false,
          "vuContractId": Array [],
        },
        Object {
          "comment": "",
          "country": "US",
          "dateSelected": null,
          "dateWithdrawn": null,
          "hide": null,
          "rightContractStatus": "MatchedOnce",
          "selected": false,
          "vuContractId": Array [],
        },
      ]
    }
    isDisabled={false}
    isRequired={false}
    name=""
    validate={[Function]}
  >
    <Component />
  </Field>
  <div
    className="nexus-c-array__add"
  >
    <WithAnalyticsContext(WithAnalyticsEvents(Button))
      className="nexus-c-dynamic-form__add-button"
      onClick={[Function]}
    >
      + Add territory
    </WithAnalyticsContext(WithAnalyticsEvents(Button))>
  </div>
  <div
    className="nexus-c-array__objects"
  >
    <div
      className="nexus-c-array__object"
      key="0"
    >
      <div
        className="nexus-c-array__remove-button"
        onClick={[Function]}
      >
        <EditorCloseIcon
          size="medium"
        />
      </div>
      <div
        className="nexus-c-array__field "
        key="nexus-c-array__field selected"
      >
        <NexusField
          allData={Object {}}
          castCrewConfig={Object {}}
          config={Array []}
          dateType=""
          dependencies={Array []}
          formData={Object {}}
          getCurrentValues={[Function]}
          getValues={[Function]}
          id="[0].selected"
          initialData={
            Object {
              "comment": "yyyyy",
              "country": "IL",
              "dateSelected": null,
              "dateWithdrawn": null,
              "hide": null,
              "rightContractStatus": "Pending",
              "selected": false,
              "vuContractId": Array [],
            }
          }
          isClearable={false}
          isEditable={true}
          isGridLayout={false}
          isHighlighted={false}
          isOptional={false}
          isReadOnly={true}
          isReadOnlyInEdit={false}
          isRequired={false}
          isRequiredVZ={false}
          isReturningTime={true}
          isTitlePage={false}
          isVerticalLayout={false}
          key="[0].selected"
          label="Selected"
          labels={Array []}
          linkConfig={Object {}}
          name="[0].selected"
          oneIsRequiredVZ={false}
          optionsConfig={Object {}}
          path="territory.selected"
          selectValues={Object {}}
          setFieldValue={[MockFunction]}
          setUpdate={[Function]}
          setUpdatedCastCrew={[Function]}
          setUpdatedValues={[Function]}
          showLocalized={false}
          tooltip={null}
          type="boolean"
          useCurrentDate={false}
          validation={Array []}
          validationError={null}
          view="EDIT"
        />
      </div>
      <div
        className="nexus-c-array__field "
        key="nexus-c-array__field country"
      >
        <NexusField
          allData={Object {}}
          castCrewConfig={Object {}}
          config={Array []}
          dateType=""
          dependencies={Array []}
          formData={Object {}}
          getCurrentValues={[Function]}
          getValues={[Function]}
          id="[0].country"
          initialData={
            Object {
              "comment": "yyyyy",
              "country": "IL",
              "dateSelected": null,
              "dateWithdrawn": null,
              "hide": null,
              "rightContractStatus": "Pending",
              "selected": false,
              "vuContractId": Array [],
            }
          }
          isClearable={false}
          isEditable={true}
          isGridLayout={false}
          isHighlighted={false}
          isOptional={false}
          isReadOnly={false}
          isReadOnlyInEdit={false}
          isRequired={false}
          isRequiredVZ={false}
          isReturningTime={true}
          isTitlePage={false}
          isVerticalLayout={false}
          key="[0].country"
          label="Country"
          labels={Array []}
          linkConfig={Object {}}
          name="[0].country"
          oneIsRequiredVZ={false}
          optionsConfig={Object {}}
          path="territory.country"
          selectValues={Object {}}
          setFieldValue={[MockFunction]}
          setUpdate={[Function]}
          setUpdatedCastCrew={[Function]}
          setUpdatedValues={[Function]}
          showLocalized={false}
          tooltip={null}
          type="string"
          useCurrentDate={false}
          validation={Array []}
          validationError={null}
          view="EDIT"
        />
      </div>
    </div>
    <div
      className="nexus-c-array__object"
      key="1"
    >
      <div
        className="nexus-c-array__remove-button"
        onClick={[Function]}
      >
        <EditorCloseIcon
          size="medium"
        />
      </div>
      <div
        className="nexus-c-array__field "
        key="nexus-c-array__field selected"
      >
        <NexusField
          allData={Object {}}
          castCrewConfig={Object {}}
          config={Array []}
          dateType=""
          dependencies={Array []}
          formData={Object {}}
          getCurrentValues={[Function]}
          getValues={[Function]}
          id="[1].selected"
          initialData={
            Object {
              "comment": "",
              "country": "US",
              "dateSelected": null,
              "dateWithdrawn": null,
              "hide": null,
              "rightContractStatus": "MatchedOnce",
              "selected": false,
              "vuContractId": Array [],
            }
          }
          isClearable={false}
          isEditable={true}
          isGridLayout={false}
          isHighlighted={false}
          isOptional={false}
          isReadOnly={true}
          isReadOnlyInEdit={false}
          isRequired={false}
          isRequiredVZ={false}
          isReturningTime={true}
          isTitlePage={false}
          isVerticalLayout={false}
          key="[1].selected"
          label="Selected"
          labels={Array []}
          linkConfig={Object {}}
          name="[1].selected"
          oneIsRequiredVZ={false}
          optionsConfig={Object {}}
          path="territory.selected"
          selectValues={Object {}}
          setFieldValue={[MockFunction]}
          setUpdate={[Function]}
          setUpdatedCastCrew={[Function]}
          setUpdatedValues={[Function]}
          showLocalized={false}
          tooltip={null}
          type="boolean"
          useCurrentDate={false}
          validation={Array []}
          validationError={null}
          view="EDIT"
        />
      </div>
      <div
        className="nexus-c-array__field "
        key="nexus-c-array__field country"
      >
        <NexusField
          allData={Object {}}
          castCrewConfig={Object {}}
          config={Array []}
          dateType=""
          dependencies={Array []}
          formData={Object {}}
          getCurrentValues={[Function]}
          getValues={[Function]}
          id="[1].country"
          initialData={
            Object {
              "comment": "",
              "country": "US",
              "dateSelected": null,
              "dateWithdrawn": null,
              "hide": null,
              "rightContractStatus": "MatchedOnce",
              "selected": false,
              "vuContractId": Array [],
            }
          }
          isClearable={false}
          isEditable={true}
          isGridLayout={false}
          isHighlighted={false}
          isOptional={false}
          isReadOnly={false}
          isReadOnlyInEdit={false}
          isRequired={false}
          isRequiredVZ={false}
          isReturningTime={true}
          isTitlePage={false}
          isVerticalLayout={false}
          key="[1].country"
          label="Country"
          labels={Array []}
          linkConfig={Object {}}
          name="[1].country"
          oneIsRequiredVZ={false}
          optionsConfig={Object {}}
          path="territory.country"
          selectValues={Object {}}
          setFieldValue={[MockFunction]}
          setUpdate={[Function]}
          setUpdatedCastCrew={[Function]}
          setUpdatedValues={[Function]}
          showLocalized={false}
          tooltip={null}
          type="string"
          useCurrentDate={false}
          validation={Array []}
          validationError={null}
          view="EDIT"
        />
      </div>
    </div>
  </div>
</div>
`;
